src/main/app/controller/courts/CourtFacilitiesController.ts

```diff
@@ -0,0 +1,125 @@
+import {AuthedRequest} from '../../../types/AuthedRequest';
+import {Response} from 'express';
+import {CSRF} from '../../../modules/csrf';
+import autobind from 'autobind-decorator';
+import {Facility, FacilityPageData, FacilityType} from '../../../types/Facility';
+import {SelectItem} from '../../../types/CourtPageData';
+import {validateDuplication} from '../../../utils/validation';
+import {Error} from '../../../types/Error';
+
+@autobind
+export class CourtFacilitiesController {
+
+  emptyNameOrDescriptionErrorMsg = 'Name and description are required for all court facilities.';
+  facilityDuplicatedErrorMsg = 'All facilities must be unique.';
+  getFacilityTypesErrorMsg = 'A problem occurred when retrieving the list of facility types.';
+  getCourtFacilitiesErrorMsg = 'A problem occurred when retrieving the court facilities.';
+  updateErrorMsg = 'A problem occurred when saving the court facilities.';
+
+  public async get(
+    req: AuthedRequest,
+    res: Response,
+    updated = false,
+    errorMsg: string[] = [],
+    courtFacilities: Facility[] = null,
+    requiresValidation = true): Promise<void> {
+    if (!courtFacilities) {
+      const slug: string = req.params.slug as string;
+      await req.scope.cradle.api.getCourtFacilities(slug)
+        .then((value: Facility[]) => courtFacilities = value)
+        .catch(() => errorMsg.push(this.getCourtFacilitiesErrorMsg));
+    }
+
+    let allFacilitiesTypes: FacilityType[] = [];
+
+    await req.scope.cradle.api.getAllFacilityTypes()
+      .then((value: FacilityType[]) => allFacilitiesTypes = value)
+      .catch(() => errorMsg.push(this.getFacilityTypesErrorMsg));
+
+    if (!courtFacilities?.some(ot => ot.isNew === true)) {
+      this.addEmptyFormsForNewEntries(courtFacilities);
+    }
+
+    const errors: Error[] = [];
+    for (const msg of errorMsg) {
+      errors.push({text: msg});
+    }
+
+    const pageData: FacilityPageData = {
+      errors: errors,
+      updated: updated,
+      facilitiesTypes: CourtFacilitiesController.getFacilityTypesForSelect(allFacilitiesTypes),
+      courtFacilities: courtFacilities,
+      requiresValidation: requiresValidation
+    };
+
+    res.render('courts/tabs/facilitiesContent', pageData);
+  }
+
+  public async put(req: AuthedRequest, res: Response): Promise<void> {
+    let courtFacilities = req.body.courtFacilities as Facility[] ?? [];
+    courtFacilities.forEach(ot => {
+      ot.isNew = (ot.isNew === true) || ((ot.isNew as any) === 'true');
+      // Workaround for the issue where the the empty row is removed after the previous name selection.
```

Comment:
is it possible to give ot.isNew a type such as String and not any?

Reply:
I updated this one. Need to convert to unknown first then convert to string. If not typescript will throw an error and says can't convert from boolean to string directly. Not sure if this new way is better.